### PR TITLE
add non linear attention following https://arxiv.org/abs/2309.08586

### PR DESCRIPTION
**Pull request by sagadre**

### CODE REVIEW

File: tests/test_relu_attention.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        [1, 1024, 2048, 8, 128],
+    ]:
+        queries = torch.rand(bs, q_seq_len, h, d)
+        keys = torch.rand(bs, k_seq_len, h, d)
+        values = torch.rand(bs, k_seq_len, h, d)
+
+        for is_causal in [True, False]:
+            torch_out = torch_attn(queries, keys, values, is_causal=is_causal)
+
+            my_out = non_linear_attn(
```
Comment (mitchellnw) [inline tests/test_relu_attention.py:20]: I like this test! What about also testing equivalence to xformers attn?

File: open_lm/attention.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            )
+            .transpose(1, 2)
+            .contiguous()
+        )
+
+
+ATTN_NON_LINEARITIES = {
+    "relu": F.relu,
+    "relu_squared": lambda x: torch.pow(F.relu(x), 2),
+    # "gelu": F.gelu, # goes to NaN with bais so comment out for now
```
Comment (mitchellnw) [inline open_lm/attention.py:76]: Not a big deal, but could explicitly just 0 out the illegal attn weights instead of adding -inf.

**Review by mitchellnw — COMMENTED** [review]
Looks great! Once we see training curves we can probably feel confident to approve/merge.

**Comment by sagadre** [comment]
curves matching for cases of xformers softmax attention and softmax attention going throught the custom attention code path
Screen Shot 2023-12-20 at 12 23 21 AM

**Comment by sagadre** [comment]
relu^2 with alpha = 1 and sequence length scalar determined by max sequence length or avg sequence length approaching the standard softmax setting. for context, this is an 11M param model trained for 220M tokens. larger runs planned for future dates, but probably not blocking this PR

Screen Shot 2023-12-20 at 11 02 51 AM

**Review by mitchellnw — APPROVED** [review]
LGTM!
